### PR TITLE
Faster ledger state

### PR DESCRIPTION
**Pull request by berewt**
Several memory/speed improvement:

- Store volatile ledger states in memory (faster rollback)
- Use Builder instead of Bytestring for serialisation
- Save LedgerState less often, and always save it on close.
- Add a `--batch-size` flag

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense and have useful messages
    - [ ] Important changes are reflected in changelog.d of the affected packages
    - [ ] Relevant tickets are mentioned in commit messages
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [ ] Targeting main unless this is a cherry-pick backport
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [ ] If relevant, reference the ADR in the PR and reference the PR in the ADR
    - [x] Reviewer requested

### CODE REVIEW

**Review by brendanrbrown — APPROVED** [review]
This all looks good. My comments are minor and mostly about documentation, particularly in cases where it's unclear that bad values of certain parameters are possible.

The CI currently fails because of some fourmolu thing.

But the tests failed for me locally because of golden files: I went ahead and fixed those up as in #286, in case you were still having issues with whitespace.

File: marconi-cardano-indexers/src/Marconi/Cardano/Indexers/EpochNonce.hs [hunk 1]
```diff
@@ -96,8 +96,8 @@ mkEpochNonceIndexer path = do
               ( epochNo INT NOT NULL
               , nonce BLOB NOT NULL
               , blockNo INT NOT NULL
-              , slotNo INT NOT NULL
-              , blockHeaderHash BLOB NOT NULL
+              , slotNo INT
+              , blockHeaderHash BLOB
```
[the remark this answers is not in the document]
Comment (berewt) [reply marconi-cardano-indexers/src/Marconi/Cardano/Indexers/EpochNonce.hs:100]: I most likely fat fingered it. I don't even remember changing it.

File: marconi-cardano-indexers/src/Marconi/Cardano/Indexers/ExtLedgerStateCoordinator.hs [hunk 2]
```diff
@@ -138,428 +122,352 @@ data EpochMetadata = EpochMetadata
...
-  , _accessToIndexer
-      :: Con.MVar
-          ( Core.WithTrace
-              IO
-              ExtLedgerStateCoordinator
-              (ExtLedgerStateEvent, WithDistance input)
-          )
-  -- ^ allow query to the indexer
+data ExtLedgerStatePersistConfig = ExtLedgerStatePersistConfig
+  { _blocksToNextSnapshot :: Word64
```
[the remark this answers is not in the document]
Comment (berewt) [reply marconi-cardano-indexers/src/Marconi/Cardano/Indexers/ExtLedgerStateCoordinator.hs:133]: I actually forgot the `StrictData` pragma at the top when I created this file a long time ago, thanks.

File: marconi-cardano-indexers/src/Marconi/Cardano/Indexers/ExtLedgerStateCoordinator.hs [hunk 3]
```diff
@@ -138,428 +122,352 @@ data EpochMetadata = EpochMetadata
...
+          Seq.Empty -> do
+            eventToSave <- Lens.use lastLedgerState
+            if isVolatile'
+              then do
+                -- we reached the first volatile event
+                -- We always save the last stable event before we reach volatile ones
+                persistConfig . blocksToNextSnapshot .= 0
+                volatileLedgerStates %= (event Seq.<|)
+              else do
+                persistConfig . blocksToNextSnapshot -= 1
```
[the remark this answers is not in the document]
Comment (berewt) [reply marconi-cardano-indexers/src/Marconi/Cardano/Indexers/ExtLedgerStateCoordinator.hs:187]: I'm adding a smart constructor and comments on the field. The other good thing is that we don't expose this type at all, so we can forge an invalid persist config now that we check the types.

**Comment by berewt** [comment]
> This all looks good. My comments are minor and mostly about documentation, particularly in cases where it's unclear that bad values of certain parameters are possible.
> 
> The CI currently fails because of some fourmolu thing.
> 
> But the tests failed for me locally because of golden files: I went ahead and fixed those up as in #286, in case you were still having issues with whitespace.

I still have these issues (I think I was able to do it in the past though, which is weird).

**Comment by brendanrbrown** [comment]
> I still have these issues

Let me know if you want me to push the whitespace fix again.

**Pull request merged**